doc/design/block_execution.md

```diff
@@ -0,0 +1,113 @@
+## Overview
+PaddlePaddle represents a neural network as a `Block`. The word block and graph are interchangable in the desgin of PaddlePaddle. A block is like a pair of curly braces in programming languages like C++ and Java, it includes some variables and a sequence of instructions, or, operators. Usually we run a block using a CPU thread. But we want to accelerate the execution of some operators using devices like GPU and FPGA.
```

Comment:
@QiJune @helinwang 
we discussed face to face and reached some agreements.
1、 To unify our discussion results, we define some common concepts to represent these modules.

- **Placement Policy**
When the operator run call happens, it needs to be assigned to a specific device. We define the  **Placement Policy** as operator assigned policy. Currently, we only need a simple priority rule to implement the simplest version.
- **Scheduler**
In the multi-node or multi-device environment, we have more than one block, need a module to manage the op running process. We define the **Scheduler** as op runner manager.
- **Converter**
The user-defined **block** is different from the one which really is executed.  There is a gap between the original block and optimized block. We define the **Converter**  as the optimizer module.

2、`Block` should not contain a `Run` interface. 
Currently, the [Block design](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/block.md#the-compilation-of-blocks) contains an interface of `RUN`, which is inherited from `OpBase`. This interface is overlapped with **Scheduler** module.

> 
```c++
 void Run(const framework::Scope& scope,
           const platform::DeviceContext& dev_ctx) const override {
    PADDLE_ENFORCE(symbols_ready_, "operators and variables should be created first.");
    for (auto& op : runtime_table_.ops()) {
      op->Run(scope, dev_ctx);
    }
  }
```
When the `block` replaces `NetOp`,  there are two concerns. 
First, the backward module use `NetOp` to represent a nested Network and a group of operators. The block contains much more than **a group of operators**, which is needed in the backward process. Second, some Op like `FC`, uses NetOp to contain more than one operator, also need to be replaced.

Reply:
Sorry I didn't catch up with your conclusions.

1. In my thoughts, placement should only describe variables, not operators. The placement is where to store the variable, and then we can know where to run the operators for sure. This is some "storage define the computes".
1. A "scheduler" often do schedule executions on limited resources, but when do multi-node or multi-device running, the nodes, CPU resources and GPU resources are already scheduled by linux kernel or kubernetes, we only need to do thread/process creating for each block and let linux or kubernetes handle them.
1.  "Converter" is definitely needed, but the name "converter" is too generic, should be named like "BlockOptimizer" or something else, what you think?

Reply:
1. One operator may have several input and output variable. We must ensure these variables are in the same device. So, we take this device as Operator's device. If one variable is not in the same device, users should add a copy operator(Paddle can do this later).
2. Maybe `Executor` is a better name?
3. Yes, `BlockOptimizer` is more specific.

Reply:
![block](https://user-images.githubusercontent.com/13348433/30842157-e18f5276-a2b2-11e7-87f0-458d4511876c.png)

This picture shows how to define variable places and add copy ops automatically.

Reply:
@typhoonzero So, where will the operator run? on gpu:0 or gpu:1. Where do We insert copy operator?before calculation or after calculation. I think we have to specify a concrete device for a operator. Please refer to the design of [TensorFlow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/node_def.proto#L48). We should have a GraphView of Block, which the data members are Nodes and Edges instead vector of VarDesc and OpDesc. The device info will located in Node in GraphView of Block.

Reply:
@typhoonzero 

> we only need to do thread/process creating for each block and let linux or kubernetes handle them.

We still need scheduling: schedule the execution of the OP only when it's ready to run (i.e. inputs are ready).

Reply:
@QiJune I see your point, I agree with defining device place for each "Node"(or operator). The reason is people define the graph by ops not by every variable, like the output of two ops is not defined by user.

When we know the variable's device place, we can easily figure out where the operator runs, this is not the reason I think.

Reply:
> If the scheduler schedules the recv OP before it's ready, it will be waiting for the send OP and block the running thread in the thread pool unnecessarily.

@helinwang Is this scheduler only for scheduling on GPU? Because when use CPU, linux kernel will put the thread to "Sleep" in this situation and do not consume CPU resource.

Reply:
@typhoonzero Sorry, my example is misleading. I mean we need a scheduler to put OP to "run queue" only when the dependencies of the OP is done. Because by definition the OP should only run when all of its dependencies is done.